pkg/virtual/generic/rootapiserver/root_apiserver.go

```diff
@@ -0,0 +1,278 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rootapiserver
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"net/url"
+	"strings"
+
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apiserver/pkg/authorization/authorizerfactory"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+	genericapiserveroptions "k8s.io/apiserver/pkg/server/options"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+
+	virtualapiserver "github.com/kcp-dev/kcp/pkg/virtual/generic/apiserver"
+	"github.com/kcp-dev/kcp/pkg/virtual/generic/builders"
+)
+
+type InformerStart func(stopCh <-chan struct{})
+type InformerStarts []InformerStart
+
+type RootAPIExtraConfig struct {
+	// we phrase it like this so we can build the post-start-hook, but no one can take more indirect dependencies on informers
+	informerStart func(stopCh <-chan struct{})
+
+	VirtualWorkspaces []builders.VirtualWorkspaceBuilder
+}
+
+// Validate helps ensure that we build this config correctly, because there are lots of bits to remember for now
+func (c *RootAPIExtraConfig) Validate() error {
+	ret := []error{}
+
+	return utilerrors.NewAggregate(ret)
+}
+
+type RootAPIConfig struct {
+	GenericConfig *genericapiserver.RecommendedConfig
+	ExtraConfig   RootAPIExtraConfig
+}
+
+// RootAPIServer is only responsible for serving the APIs for the virtual workspace
+// at a given root path or root path family
+// It does NOT expose oauth, related oauth endpoints, or any kube APIs.
+type RootAPIServer struct {
+	GenericAPIServer *genericapiserver.GenericAPIServer
+}
+
+type completedConfig struct {
+	GenericConfig genericapiserver.CompletedConfig
+	ExtraConfig   *RootAPIExtraConfig
+}
+
+type CompletedConfig struct {
+	// Embed a private pointer that cannot be instantiated outside of this package.
+	*completedConfig
+}
+
+// Complete fills in any fields not set that are required to have valid data. It's mutating the receiver.
+func (c *RootAPIConfig) Complete() completedConfig {
+	cfg := completedConfig{
+		c.GenericConfig.Complete(),
+		&c.ExtraConfig,
+	}
+
+	return cfg
+}
+
+func (c *completedConfig) withAPIServerForAPIGroup(virtualWorkspaceName string, groupAPIServerBuilder builders.APIGroupAPIServerBuilder) apiServerAppenderFunc {
+	return func(delegateAPIServer genericapiserver.DelegationTarget) (genericapiserver.DelegationTarget, error) {
+		restStorageBuilders, err := groupAPIServerBuilder.Initialize(c.GenericConfig)
+		cfg := &virtualapiserver.GroupAPIServerConfig{
+			GenericConfig: &genericapiserver.RecommendedConfig{Config: *c.GenericConfig.Config, SharedInformerFactory: c.GenericConfig.SharedInformerFactory},
+			ExtraConfig: virtualapiserver.ExtraConfig{
+				Codecs:          legacyscheme.Codecs,
+				Scheme:          legacyscheme.Scheme,
```

Comment:
we should not do this. It will be hell to remove later.

Reply:
Fixed in https://github.com/davidfestal/kcp/pull/2

Reply:
Fixed by cherry picking some changes in your PR commits